build-system/tasks/amp-task-runner.js

```diff
@@ -0,0 +1,153 @@
+ */
+async function runTask(taskName, taskFunc) {
+  const taskFile = '~' + path.sep + path.relative(os.homedir(), 'amp.js');
+  log('Using task file', magenta(taskFile));
+  const start = Date.now();
+  try {
+    log(`Starting '${cyan(taskName)}'...`);
+    await taskFunc();
+    log('Finished', `'${cyan(taskName)}'`, 'after', magenta(getTime(start)));
+  } catch (err) {
+    log(cyan(taskName), red('errored after'), magenta(getTime(start)));
+    log(err);
+    process.exitCode = 1;
+  }
+}
+
+/**
+ * Helper that creates the tasks in AMP's toolchain based on the invocation:
+ * - For `amp --help`, load all task descriptions so a list can be printed.
+ * - For other tasks, load the entry point, validate usage, and run the task.
+ * @param {string} taskName
+ * @param {string} taskFuncName
+ * @param {string} taskSourceFileName
+ */
+function createTask(taskName, taskFuncName, taskSourceFileName) {
+  const taskSourceFilePath = `./${taskSourceFileName}`;
+  const isInvokedTask = argv._.includes(taskName); // `amp <task>`
+  const isDefaultTask =
+    argv._.length == 0 && taskName == 'default' && !isHelpTask; // `amp`
+
+  if (isHelpTask) {
+    const taskFunc = require(taskSourceFilePath)[taskFuncName];
+    const task = commander.command(cyan(taskName));
+    task.description(taskFunc.description);
+  }
+  if (isInvokedTask || isDefaultTask) {
+    const taskFunc = require(taskSourceFilePath)[taskFuncName];
+    const task = commander.command(taskName, {isDefault: isDefaultTask});
+    task.description(green(taskFunc.description));
+    task.allowUnknownOption(); // Fall through to validateUsage()
+    task.helpOption('--help', 'Print this list of flags');
+    task.usage('<flags>');
+    for (const [flag, description] of Object.entries(taskFunc.flags ?? {})) {
+      task.option(`--${cyan(flag)}`, description);
+    }
+    task.action(async () => {
+      validateUsage(task, taskName, taskFunc);
+      await runTask(taskName, taskFunc);
+    });
+  }
+}
+
+/**
+ * Validates usage by examining task and flag invocation.
+ * @param {Object} task
+ * @param {string} taskName
+ * @param {function} taskFunc
+ */
+function validateUsage(task, taskName, taskFunc) {
+  const tasks = argv._;
+  const invalidTasks = tasks.filter((task) => task != taskName);
+
+  const flags = Object.keys(argv).slice(1); // Everything after '_'
+  const validFlags = taskFunc.flags ? Object.keys(taskFunc.flags) : [];
+  const invalidFlags = flags.filter((flag) => !validFlags.includes(flag));
+
+  if (invalidTasks.length > 0 || invalidFlags.length > 0) {
+    task.addHelpText('before', red('ERROR: ') + 'Invalid usage');
+    task.help({error: true});
+  }
+}
+
+/**
+ * Finalizes the task runner by doing special-case setup for `amp --help`,
+ * parsing the invoked command, and printing an error message if an unknown task
+ * was called.
+ */
+function finalizeRunner() {
+  commander.addHelpCommand(false); // We already have `amp --help` and `amp <task> --help`
```

Comment:
Is there a way to just implement this so that the default behaviour of `commander` is used, instead of reinventing it?

Reply:
Commander provides two implementations of help by default. In our case, they are `amp help` and `amp --help`. After lots of testing / trial and error, `amp --help` was more intuitive. (`amp help` made it hard to tell if `help` is a task name or something else.) All we're doing here is preventing the addition of yet another way of doing things.